### PR TITLE
Add some missing operator<< to runtime library.

### LINKED ISSUE #159: Flow-tracing generates invalid code
Compiling Spicy code with `-X flow` produces C++ code that's not going to compile. This has been the case for a while, just recording here for now. (`-X trace` works fine).

### PR DESCRIPTION
**Pull request by rsmmr**
This now lets "-X trace" work for Spocy parsers.

Closes #159.

### CODE REVIEW

**Comment by rsmmr** [comment]
@bbannier have you looked at this yet?

File: hilti/runtime/include/types/optional.h [hunk 1]
```diff
@@ -75,3 +75,14 @@ inline std::string detail::to_string_for_print<std::optional<std::string_view>>(
 }
 
 } // namespace hilti::rt
+
+namespace std {
+
+// Provide operator<< overload for optionals that have a custom HILTI-side to_string() conversion.
+template<typename T>
+inline auto operator<<(std::ostream& out, const std::optional<T>& x) -> decltype(hilti::rt::detail::adl::to_string(x), out) {
```
Comment (bbannier) [inline hilti/runtime/include/types/optional.h:83]: Let's cover this with a simple test in `hilti/runtime/src/tests/to_string.cc`.

File: spicy/runtime/include/parsed-unit.h [hunk 2]
```diff
@@ -72,3 +73,11 @@ class ParsedUnit : public hilti::rt::type_info::value::Parent {
 };
 
 } // namespace spicy::rt
+
+namespace hilti::rt::detail::adl {
+inline std::string to_string(const ::spicy::rt::ParsedUnit& u, adl::tag /*unused*/) { return "<parsed unit>"; };
```
Comment (bbannier) [inline spicy/runtime/include/parsed-unit.h:78]: Ditto.

**Review by bbannier — APPROVED** [review]

File: hilti/runtime/src/tests/to_string.cc [hunk 3]
```diff
@@ -270,4 +272,16 @@ TEST_CASE("View") {
              "G\\xc3\\xa4nsef\\xc3\\xbc\\xc3\\x9fchen\\x00\\x01\\x02");
 }
 
+TEST_CASE("Optional") {
```
Comment (bbannier) [inline hilti/runtime/src/tests/to_string.cc:275]: Let's group this with `TEST_CASE("optional")` down around l.136.

**Review by bbannier — APPROVED** [review]

**Pull request merged**